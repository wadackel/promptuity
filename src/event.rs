//! A module to read events.
//!
//! # Event
//!
//! The `event` module provides functionalities used in prompt handlers.
//! The main offering is the direct re-export of [`crossterm::event`].
//!
//! ## Examples
//!
//! ```ignore
//! use promptuity::event::{KeyCode, KeyModifiers};
//!
//! impl<W: std::io::Write> Prompt<W> for YourPrompt {
//!   // ...
//!   fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> PromptState {
//!     match (code, modifiers) {
//!       (KeyCode::Enter, KeyModifiers::NONE) => PromptState::Submit,
//!       _ => PromptState::Active,
//!     }
//!   }
//!   // ...
//! }
//! ```

/// [`KeyCode`] re-exports from [`crossterm::event`].
pub use crossterm::event::KeyCode;
/// [`KeyModifiers`] re-exports from [`crossterm::event`].
pub use crossterm::event::KeyModifiers;
