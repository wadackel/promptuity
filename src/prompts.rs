//! A module that provides built-in prompts.
//!
//! # Prompts
//!
//! The `prompts` module provides the following built-in prompts.  
//! These built-in prompts can also be used as reference implementations when building your custom prompts.
//!
//! - [`Input`]: A prompt for general text input.
//! - [`Password`]: A text input prompt where the input is not displayed.
//! - [`Number`]: A prompt for inputting only integer values.
//! - [`Select`]: A prompt for selecting a single element from a list of options.
//! - [`MultiSelect`]: A prompt for selecting multiple elements from a list of options.
//! - [`Confirm`]: A prompt for inputting a Yes/No choice.
//!
//! # Examples
//!
//! This is an example of using a simple built-in prompt.
//!
//! ```no_run
//! use promptuity::themes::MinimalTheme;
//! use promptuity::prompts::{Input, Confirm};
//! use promptuity::{Promptuity, Term};
//!
//! # fn main() -> Result<(), promptuity::Error> {
//! let mut term = Term::default();
//! let mut theme = MinimalTheme::default();
//! let mut p = Promptuity::new(&mut term, &mut theme);
//!
//! p.begin()?;
//!
//! let name = p.prompt(Input::new("Please enter your username").with_placeholder("username"))?;
//! let full = p.prompt(Confirm::new("Are you a full-time software developer?").with_default(true))?;
//!
//! p.finish()?;
//! # Ok(())
//! # }
//! ```

mod confirm;
mod input;
mod multi_select;
mod number;
mod password;
mod select;
#[cfg(test)]
pub(crate) mod test;

pub use confirm::*;
pub use input::*;
pub use multi_select::*;
pub use number::*;
pub use password::*;
pub use select::*;
