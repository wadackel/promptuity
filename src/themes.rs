//! A module that provides built-in themes.
//!
//! # Themes
//!
//! The `themes` module provides the following built-in themes.
//!
//! - [`MinimalTheme`]: A Theme that offers a compact and minimalistic display.
//! - [`FancyTheme`]: A Theme that displays with a rich UI.

mod fancy;
mod minimal;

pub use fancy::*;
pub use minimal::*;
