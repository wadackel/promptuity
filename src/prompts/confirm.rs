use crate::event::*;
use crate::style::{Color, Styled, Symbol};
use crate::{Prompt, PromptInput, PromptState, RenderPayload};

const S_ACTIVE: Symbol = Symbol("●", ">");
const S_INACTIVE: Symbol = Symbol("○", " ");

/// A trait for formatting the [`Confirm`] prompt.
///
/// All methods have default implementations, allowing you to override only the specific formatting process you need.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::{Confirm, ConfirmFormatter};
///
/// struct CustomFormatter;
///
/// impl ConfirmFormatter for CustomFormatter {
///     fn layout(&self, yes: String, no: String) -> String {
///         format!("{} or {}", yes, no)
///     }
/// }
///
/// let _ = Confirm::new("...").with_formatter(CustomFormatter);
/// ```
pub trait ConfirmFormatter {
    /// Formats the "Yes" option.
    fn yes(&self, active: bool) -> String {
        let icon = if active {
            Styled::new(S_ACTIVE).fg(Color::Green).to_string()
        } else {
            Styled::new(S_INACTIVE).fg(Color::DarkGrey).to_string()
        };
        format!("{} Yes", icon)
    }

    /// Formats the "No" option.
    fn no(&self, active: bool) -> String {
        let icon = if active {
            Styled::new(S_ACTIVE).fg(Color::Green).to_string()
        } else {
            Styled::new(S_INACTIVE).fg(Color::DarkGrey).to_string()
        };
        format!("{} No", icon)
    }

    /// Formats the layout of the active prompt.
    fn layout(&self, yes: String, no: String) -> String {
        format!("{}  /  {}", yes, no)
    }

    /// Formats the submitted value.
    fn submit(&self, value: bool) -> String {
        if value {
            "Yes".into()
        } else {
            "No".into()
        }
    }
}

/// The default formatter for [`Confirm`].
pub struct DefaultConfirmFormatter;

impl DefaultConfirmFormatter {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self
    }
}

impl ConfirmFormatter for DefaultConfirmFormatter {}

/// A prompt for inputting a Yes/No choice.
///
/// # Options
///
/// - **Formatter**: Customizes the prompt display. See [`ConfirmFormatter`].
/// - **Hint**: A message to assist with field input. Defaults to `None`.
/// - **Default Value**: The default value of `bool`. Defaults to `false`.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::Confirm;
///
/// let _ = Confirm::new("Are you sure?").with_default(true);
/// ```
pub struct Confirm {
    formatter: Box<dyn ConfirmFormatter>,
    message: String,
    hint: Option<String>,
    value: bool,
}

impl Confirm {
    /// Creates a new [`Confirm`] prompt.
    pub fn new(message: impl std::fmt::Display) -> Self {
        Self {
            formatter: Box::new(DefaultConfirmFormatter),
            message: message.to_string(),
            hint: None,
            value: false,
        }
    }

    /// Sets the formatter for the prompt.
    pub fn with_formatter(&mut self, formatter: impl ConfirmFormatter + 'static) -> &mut Self {
        self.formatter = Box::new(formatter);
        self
    }

    /// Sets the hint message for the prompt.
    pub fn with_hint(&mut self, hint: impl std::fmt::Display) -> &mut Self {
        self.hint = Some(hint.to_string());
        self
    }

    /// Sets the default value for the prompt.
    pub fn with_default(&mut self, value: bool) -> &mut Self {
        self.value = value;
        self
    }
}

impl AsMut<Confirm> for Confirm {
    fn as_mut(&mut self) -> &mut Self {
        self
    }
}

impl Prompt for Confirm {
    type Output = bool;

    fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> PromptState {
        match (code, modifiers) {
            (KeyCode::Enter, _) => PromptState::Submit,
            (KeyCode::Esc, _) | (KeyCode::Char('c'), KeyModifiers::CONTROL) => PromptState::Cancel,
            (KeyCode::Char('y'), KeyModifiers::NONE) | (KeyCode::Char('Y'), KeyModifiers::NONE) => {
                self.value = true;
                PromptState::Submit
            }
            (KeyCode::Char('n'), KeyModifiers::NONE) | (KeyCode::Char('N'), KeyModifiers::NONE) => {
                self.value = false;
                PromptState::Submit
            }
            (KeyCode::Left, _)
            | (KeyCode::Char('h'), _)
            | (KeyCode::Char('p'), KeyModifiers::CONTROL) => {
                self.value = true;
                PromptState::Active
            }
            (KeyCode::Right, _)
            | (KeyCode::Char('l'), _)
            | (KeyCode::Char('n'), KeyModifiers::CONTROL) => {
                self.value = false;
                PromptState::Active
            }
            _ => PromptState::Active,
        }
    }

    fn render(&mut self, state: &PromptState) -> Result<RenderPayload, String> {
        let payload = RenderPayload::new(self.message.clone(), self.hint.clone(), None);

        match state {
            PromptState::Submit => {
                Ok(payload.input(PromptInput::Raw(self.formatter.submit(self.value))))
            }

            PromptState::Cancel => Ok(payload),

            _ => Ok(payload.input(PromptInput::Raw(self.formatter.layout(
                self.formatter.yes(self.value),
                self.formatter.no(!self.value),
            )))),
        }
    }

    fn submit(&mut self) -> Self::Output {
        self.value
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_prompt;

    test_prompt!(
        test_hint,
        Confirm::new("test message").with_hint("hint message"),
        vec![]
    );

    test_prompt!(
        test_default_enter,
        Confirm::new("test message").as_mut(),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_default_yes,
        Confirm::new("test message").with_default(true),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_move,
        Confirm::new("test message").as_mut(),
        vec![
            (KeyCode::Left, KeyModifiers::NONE),
            (KeyCode::Right, KeyModifiers::NONE),
            (KeyCode::Char('h'), KeyModifiers::NONE),
            (KeyCode::Char('l'), KeyModifiers::NONE),
            (KeyCode::Char('h'), KeyModifiers::CONTROL),
            (KeyCode::Char('l'), KeyModifiers::CONTROL),
            (KeyCode::Char('p'), KeyModifiers::CONTROL),
            (KeyCode::Char('n'), KeyModifiers::CONTROL),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );

    test_prompt!(
        test_direct_lower_yes,
        Confirm::new("test message").with_default(false),
        vec![(KeyCode::Char('y'), KeyModifiers::NONE)]
    );

    test_prompt!(
        test_direct_upper_yes,
        Confirm::new("test message").with_default(false),
        vec![(KeyCode::Char('Y'), KeyModifiers::NONE)]
    );

    test_prompt!(
        test_direct_lower_no,
        Confirm::new("test message").with_default(true),
        vec![(KeyCode::Char('n'), KeyModifiers::NONE)]
    );

    test_prompt!(
        test_direct_upper_no,
        Confirm::new("test message").with_default(true),
        vec![(KeyCode::Char('N'), KeyModifiers::NONE)]
    );
}
