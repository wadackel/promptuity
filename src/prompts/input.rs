use crate::event::*;
use crate::{InputCursor, Prompt, PromptInput, PromptState, RenderPayload, Validator};

/// A trait for formatting the [`Input`] prompt.
///
/// All methods have default implementations, allowing you to override only the specific formatting process you need.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::{Input, InputFormatter};
///
/// struct CustomFormatter;
///
/// impl InputFormatter for CustomFormatter {
///     fn err_required(&self) -> String {
///         "REQUIRED!!".into()
///     }
/// }
///
/// let _ = Input::new("...").with_formatter(CustomFormatter);
/// ```
pub trait InputFormatter {
    /// Formats the error message when the input is empty and required.
    fn err_required(&self) -> String {
        "This field is required.".into()
    }
}

/// The default formatter for [`Input`].
pub struct DefaultInputFormatter;

impl DefaultInputFormatter {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self {}
    }
}

impl InputFormatter for DefaultInputFormatter {}

/// A prompt for general text input.
///
/// # Options
///
/// - **Formatter**: Customizes the prompt display. See [`InputFormatter`].
/// - **Hint**: A message to assist with field input. Defaults to `None`.
/// - **Placeholder**: An auxiliary message displayed when no input is given.
/// - **Required**: A flag indicating whether to allow no input.
/// - **Default Value**: The default value of `String`.
/// - **Validator**: A function to validate the value at the time of submission.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::Input;
///
/// let _ = Input::new("What is your accout name?").with_hint("e.g. wadackel");
/// ```
pub struct Input {
    formatter: Box<dyn InputFormatter>,
    message: String,
    hint: Option<String>,
    placeholder: Option<String>,
    required: bool,
    validator: Option<Box<dyn Validator<String>>>,
    input: InputCursor,
}

impl Input {
    /// Creates a new [`Input`] prompt.
    pub fn new(message: impl std::fmt::Display) -> Self {
        Self {
            formatter: Box::new(DefaultInputFormatter::new()),
            message: message.to_string(),
            hint: None,
            placeholder: None,
            required: true,
            validator: None,
            input: InputCursor::default(),
        }
    }

    /// Sets the formatter for the prompt.
    pub fn with_formatter(&mut self, formatter: impl InputFormatter + 'static) -> &mut Self {
        self.formatter = Box::new(formatter);
        self
    }

    /// Sets the hint message for the prompt.
    pub fn with_hint(&mut self, hint: impl std::fmt::Display) -> &mut Self {
        self.hint = Some(hint.to_string());
        self
    }

    /// Sets the placeholder message for the prompt.
    pub fn with_placeholder(&mut self, placeholder: impl std::fmt::Display) -> &mut Self {
        self.placeholder = Some(placeholder.to_string());
        self
    }

    /// Sets the required flag for the prompt.
    pub fn with_required(&mut self, required: bool) -> &mut Self {
        self.required = required;
        self
    }

    /// Sets the default value for the prompt.
    pub fn with_default(&mut self, value: impl std::fmt::Display) -> &mut Self {
        self.input = InputCursor::from(value.to_string());
        self
    }

    /// Sets the validator for the prompt.
    pub fn with_validator(&mut self, f: impl Validator<String> + 'static) -> &mut Self {
        self.validator = Some(Box::new(move |value: &String| -> Result<(), String> {
            f.validate(value).map_err(|err| err.to_string())
        }));
        self
    }
}

impl AsMut<Input> for Input {
    fn as_mut(&mut self) -> &mut Self {
        self
    }
}

impl<W: std::io::Write> Prompt<W> for Input {
    type Output = String;

    fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> PromptState {
        match (code, modifiers) {
            (KeyCode::Esc, _) | (KeyCode::Char('c'), KeyModifiers::CONTROL) => PromptState::Cancel,
            (KeyCode::Enter, _) => {
                if self.input.is_empty() && self.required {
                    PromptState::Error(self.formatter.err_required())
                } else {
                    PromptState::Submit
                }
            }
            (KeyCode::Left, _) | (KeyCode::Char('b'), KeyModifiers::CONTROL) => {
                self.input.move_left();
                PromptState::Active
            }
            (KeyCode::Right, _) | (KeyCode::Char('f'), KeyModifiers::CONTROL) => {
                self.input.move_right();
                PromptState::Active
            }
            (KeyCode::Home, _) | (KeyCode::Char('a'), KeyModifiers::CONTROL) => {
                self.input.move_home();
                PromptState::Active
            }
            (KeyCode::End, _) | (KeyCode::Char('e'), KeyModifiers::CONTROL) => {
                self.input.move_end();
                PromptState::Active
            }
            (KeyCode::Backspace, _) | (KeyCode::Char('h'), KeyModifiers::CONTROL) => {
                self.input.delete_left_char();
                PromptState::Active
            }
            (KeyCode::Char('w'), KeyModifiers::CONTROL) => {
                self.input.delete_left_word();
                PromptState::Active
            }
            (KeyCode::Delete, _) | (KeyCode::Char('d'), KeyModifiers::CONTROL) => {
                self.input.delete_right_char();
                PromptState::Active
            }
            (KeyCode::Char('k'), KeyModifiers::CONTROL) => {
                self.input.delete_rest_line();
                PromptState::Active
            }
            (KeyCode::Char('u'), KeyModifiers::CONTROL) => {
                self.input.delete_line();
                PromptState::Active
            }
            (KeyCode::Char(c), _) => {
                self.input.insert(c);
                PromptState::Active
            }
            _ => PromptState::Active,
        }
    }

    fn submit(&mut self) -> Self::Output {
        self.input.value()
    }

    fn render(&mut self, _: &PromptState) -> Result<RenderPayload, String> {
        Ok(RenderPayload::new(
            self.message.clone(),
            self.hint.clone(),
            self.placeholder.clone(),
        )
        .input(PromptInput::Cursor(self.input.clone())))
    }

    fn validate(&self) -> Result<(), String> {
        self.validator
            .as_ref()
            .map_or(Ok(()), |validator| validator.validate(&self.input.value()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_prompt;

    test_prompt!(
        test_hint,
        Input,
        Input::new("test message").with_hint("hint message"),
        vec![]
    );

    test_prompt!(
        test_placeholder,
        Input,
        Input::new("test message").with_placeholder("placeholder message"),
        vec![]
    );

    test_prompt!(
        test_default,
        Input,
        Input::new("test message").as_mut(),
        vec![
            (KeyCode::Char('a'), KeyModifiers::NONE),
            (KeyCode::Char('b'), KeyModifiers::NONE),
            (KeyCode::Char('c'), KeyModifiers::NONE),
            (KeyCode::Left, KeyModifiers::NONE),
            (KeyCode::Left, KeyModifiers::NONE),
            (KeyCode::Left, KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );

    test_prompt!(
        test_required_error,
        Input,
        Input::new("test message").as_mut(),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_non_required_empty_submit,
        Input,
        Input::new("test message").with_required(false),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_move,
        Input,
        Input::new("test message").with_default("abcdef"),
        vec![
            (KeyCode::Left, KeyModifiers::NONE),
            (KeyCode::Char('b'), KeyModifiers::CONTROL),
            (KeyCode::Home, KeyModifiers::NONE),
            (KeyCode::Right, KeyModifiers::NONE),
            (KeyCode::Char('f'), KeyModifiers::CONTROL),
            (KeyCode::End, KeyModifiers::NONE),
            (KeyCode::Char('a'), KeyModifiers::CONTROL),
        ]
    );

    test_prompt!(
        test_editing,
        Input,
        Input::new("test message").with_default("abcdef"),
        vec![
            (KeyCode::Backspace, KeyModifiers::NONE),
            (KeyCode::Char('h'), KeyModifiers::CONTROL),
            (KeyCode::Char('h'), KeyModifiers::NONE),
            (KeyCode::Home, KeyModifiers::NONE),
            (KeyCode::Delete, KeyModifiers::NONE),
            (KeyCode::Right, KeyModifiers::NONE),
            (KeyCode::Char('d'), KeyModifiers::CONTROL),
            (KeyCode::Char('k'), KeyModifiers::CONTROL),
            (KeyCode::Char('a'), KeyModifiers::NONE),
            (KeyCode::Char('r'), KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Char('b'), KeyModifiers::NONE),
            (KeyCode::Char('a'), KeyModifiers::NONE),
            (KeyCode::Char('z'), KeyModifiers::NONE),
            (KeyCode::Char('w'), KeyModifiers::CONTROL),
            (KeyCode::Char('w'), KeyModifiers::CONTROL),
            (KeyCode::Char('a'), KeyModifiers::NONE),
            (KeyCode::Char('b'), KeyModifiers::NONE),
            (KeyCode::Char('c'), KeyModifiers::NONE),
            (KeyCode::Left, KeyModifiers::NONE),
            (KeyCode::Left, KeyModifiers::NONE),
            (KeyCode::Char('u'), KeyModifiers::CONTROL),
        ]
    );

    test_prompt!(
        test_validation,
        Input,
        Input::new("test message").with_validator(|v: &String| {
            if v.as_str() == "abc" {
                Err("Error Message".into())
            } else {
                Ok(())
            }
        }),
        vec![
            (KeyCode::Char('a'), KeyModifiers::NONE),
            (KeyCode::Char('b'), KeyModifiers::NONE),
            (KeyCode::Char('c'), KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
            (KeyCode::Backspace, KeyModifiers::NONE),
            (KeyCode::Char('z'), KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );
}
