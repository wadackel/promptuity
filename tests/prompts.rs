use pretty_assertions::assert_eq;
use promptuity::event::*;
use promptuity::prompts::{
    Confirm, Input, MultiSelect, MultiSelectOption, Number, Select, SelectOption,
};
use promptuity::themes::MinimalTheme;
use promptuity::Promptuity;

mod fake_term;

#[derive(Debug, PartialEq)]
struct TestResult {
    input: String,
    password: String,
    number: isize,
    select: usize,
    multi_select: Vec<usize>,
    confirm: bool,
}

#[test]
fn test_prompts() {
    let mut term = fake_term::Term::new(&[
        // Input
        (KeyCode::Char('a'), KeyModifiers::NONE),
        (KeyCode::Char('b'), KeyModifiers::NONE),
        (KeyCode::Char('c'), KeyModifiers::NONE),
        (KeyCode::Enter, KeyModifiers::NONE),
        // Password
        (KeyCode::Char('f'), KeyModifiers::NONE),
        (KeyCode::Char('o'), KeyModifiers::NONE),
        (KeyCode::Char('o'), KeyModifiers::NONE),
        (KeyCode::Enter, KeyModifiers::NONE),
        // Number
        (KeyCode::Char('1'), KeyModifiers::NONE),
        (KeyCode::Char('0'), KeyModifiers::NONE),
        (KeyCode::Char('0'), KeyModifiers::NONE),
        (KeyCode::Home, KeyModifiers::NONE),
        (KeyCode::Char('-'), KeyModifiers::NONE),
        (KeyCode::Enter, KeyModifiers::NONE),
        // Select
        (KeyCode::Char('j'), KeyModifiers::NONE),
        (KeyCode::Down, KeyModifiers::NONE),
        (KeyCode::Enter, KeyModifiers::NONE),
        // MultiSelect
        (KeyCode::Char(' '), KeyModifiers::NONE),
        (KeyCode::Down, KeyModifiers::NONE),
        (KeyCode::Char('j'), KeyModifiers::NONE),
        (KeyCode::Char(' '), KeyModifiers::NONE),
        (KeyCode::Enter, KeyModifiers::NONE),
        // Confirm
        (KeyCode::Right, KeyModifiers::NONE),
        (KeyCode::Left, KeyModifiers::NONE),
        (KeyCode::Char('n'), KeyModifiers::NONE),
        (KeyCode::Enter, KeyModifiers::NONE),
    ]);

    let mut theme = MinimalTheme::default();

    let result = {
        let mut p = Promptuity::new(&mut term, &mut theme);
        let input = p.prompt(Input::new("Input Message").as_mut()).unwrap();
        let password = p.prompt(Input::new("Password Message").as_mut()).unwrap();
        let number = p.prompt(Number::new("Number Message").as_mut()).unwrap();
        let select = p
            .prompt(
                Select::new(
                    "Select Message",
                    vec![
                        SelectOption::new("Option1", 1),
                        SelectOption::new("Option2", 2),
                        SelectOption::new("Option3", 3),
                    ],
                )
                .as_mut(),
            )
            .unwrap();
        let multi_select = p
            .prompt(
                MultiSelect::new(
                    "MultiSelect Message",
                    vec![
                        MultiSelectOption::new("Option1", 1),
                        MultiSelectOption::new("Option2", 2),
                        MultiSelectOption::new("Option3", 3),
                    ],
                )
                .as_mut(),
            )
            .unwrap();
        let confirm = p.prompt(Confirm::new("Confirm Message").as_mut()).unwrap();
        TestResult {
            input,
            password,
            number,
            select,
            multi_select,
            confirm,
        }
    };

    let output = term.output();

    assert_eq!(
        TestResult {
            input: "abc".into(),
            password: "foo".into(),
            number: -100,
            select: 3,
            multi_select: vec![1, 3],
            confirm: false,
        },
        result
    );

    insta::with_settings!({ omit_expression => true }, {
        insta::assert_snapshot!(output);
    });
}
