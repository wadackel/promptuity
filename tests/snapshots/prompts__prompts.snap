---
source: tests/prompts.rs
---
[38;5;14m?[39m [1mInput Message[0m  [7m [0m[38;5;8m[39m
[38;5;14m?[39m [1mInput Message[0m  a[7m [0m
[38;5;14m?[39m [1mInput Message[0m  ab[7m [0m
[38;5;14m?[39m [1mInput Message[0m  abc[7m [0m
[38;5;10m‚úî[39m [1mInput Message[0m  [38;5;14mabc[39m
[38;5;14m?[39m [1mPassword Message[0m  [7m [0m[38;5;8m[39m
[38;5;14m?[39m [1mPassword Message[0m  f[7m [0m
[38;5;14m?[39m [1mPassword Message[0m  fo[7m [0m
[38;5;14m?[39m [1mPassword Message[0m  foo[7m [0m
[38;5;10m‚úî[39m [1mPassword Message[0m  [38;5;14mfoo[39m
[38;5;14m?[39m [1mNumber Message[0m  [7m [0m[38;5;8m[39m
[38;5;14m?[39m [1mNumber Message[0m  1[7m [0m
[38;5;14m?[39m [1mNumber Message[0m  10[7m [0m
[38;5;14m?[39m [1mNumber Message[0m  100[7m [0m
[38;5;14m?[39m [1mNumber Message[0m  [7m1[0m00
[38;5;14m?[39m [1mNumber Message[0m  -[7m1[0m00
[38;5;10m‚úî[39m [1mNumber Message[0m  [38;5;14m-100[39m
[38;5;14m?[39m [1mSelect Message[0m
[38;5;10m‚óâ[39m [4mOption1[0m
[38;5;8m‚óØ[39m [38;5;8mOption2[39m
[38;5;8m‚óØ[39m [38;5;8mOption3[39m
[38;5;14m?[39m [1mSelect Message[0m
[38;5;8m‚óØ[39m [38;5;8mOption1[39m
[38;5;10m‚óâ[39m [4mOption2[0m
[38;5;8m‚óØ[39m [38;5;8mOption3[39m
[38;5;14m?[39m [1mSelect Message[0m
[38;5;8m‚óØ[39m [38;5;8mOption1[39m
[38;5;8m‚óØ[39m [38;5;8mOption2[39m
[38;5;10m‚óâ[39m [4mOption3[0m
[38;5;10m‚úî[39m [1mSelect Message[0m  [38;5;14mOption3[39m
[38;5;14m?[39m [1mMultiSelect Message[0m
[38;5;8m‚óØ[39m [4mOption1[0m
[38;5;8m‚óØ[39m [38;5;8mOption2[39m
[38;5;8m‚óØ[39m [38;5;8mOption3[39m
[38;5;14m?[39m [1mMultiSelect Message[0m
[38;5;10m‚óâ[39m [4mOption1[0m
[38;5;8m‚óØ[39m [38;5;8mOption2[39m
[38;5;8m‚óØ[39m [38;5;8mOption3[39m
[38;5;14m?[39m [1mMultiSelect Message[0m
[38;5;10m‚óâ[39m [38;5;8mOption1[39m
[38;5;8m‚óØ[39m [4mOption2[0m
[38;5;8m‚óØ[39m [38;5;8mOption3[39m
[38;5;14m?[39m [1mMultiSelect Message[0m
[38;5;10m‚óâ[39m [38;5;8mOption1[39m
[38;5;8m‚óØ[39m [38;5;8mOption2[39m
[38;5;8m‚óØ[39m [4mOption3[0m
[38;5;14m?[39m [1mMultiSelect Message[0m
[38;5;10m‚óâ[39m [38;5;8mOption1[39m
[38;5;8m‚óØ[39m [38;5;8mOption2[39m
[38;5;10m‚óâ[39m [4mOption3[0m
[38;5;10m‚úî[39m [1mMultiSelect Message[0m  [38;5;14mOption1, Option3[39m
[38;5;14m?[39m [1mConfirm Message[0m  [38;5;8m‚óã[39m Yes  /  [38;5;10m‚óè[39m No
[38;5;14m?[39m [1mConfirm Message[0m  [38;5;8m‚óã[39m Yes  /  [38;5;10m‚óè[39m No
[38;5;14m?[39m [1mConfirm Message[0m  [38;5;10m‚óè[39m Yes  /  [38;5;8m‚óã[39m No
[38;5;10m‚úî[39m [1mConfirm Message[0m  [38;5;14mNo[39m

